### PR TITLE
BPI

### PR DESCRIPTION
**Pull request by Panakotta00**
Merges the current state of BPI to the Dev branch containing:
- better UClass impl.
- better UObject impl
- better UProperty impl
- better UFunction impl
- better FString impl
- better TArray impl
- better FWeakObjectPointer impl
- better delegate impl
- add UProperty Builder
- add UFunction Builder
- UFunction injection to existing UClasses
- add UClass Builder
- add UEnum Builder
& more

### CODE REVIEW

File: SatisfactoryModLoader/util/Utility.cpp [hunk 1]
```diff
@@ -5,10 +5,57 @@
...
+			auto GObjectsOffset = *reinterpret_cast<uint32_t*>(GObjectsAddress + 3);
+			Objects::UObject::objs = reinterpret_cast<Objects::FUObjectArray*>(GObjectsAddress + 7 + GObjectsOffset);
+
+			FName::fNameConstruct_f = (void (WINAPI*)(void*, const wchar_t*, FName::EFindName)) DetourFindFunction("FactoryGame-Win64-Shipping.exe", "FName::FName");
+			UClass::findFunction_f = (UFunction* (WINAPI*)(UObject*, FName)) DetourFindFunction("FactoryGame-Win64-Shipping.exe", "UObject::FindFunction");
+		
+			FMemory::init();
+		}
+	}
+
```
Comment (Vilsol) [inline SatisfactoryModLoader/util/Utility.cpp]: Why are you duplicating already existing functions?

https://github.com/satisfactorymodding/SatisfactoryModLoader/blob/master/SatisfactorySDK/SDK.hpp#L1790-L1835

File: SatisfactoryModLoader/assets/BPInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+		void(*privateStaticClassBody)(const TCHAR*, const TCHAR*, Objects::UClass*&, void(*)(), std::uint32_t, Objects::EClassFlags, Objects::EClassCastFlags, const TCHAR*, void(*)(FObjectInitializer&), Objects::UObject*(*)(FVtableHelper&), void(*)(Objects::UObject*, FReferenceCollector&), Objects::UClass*(*)(), Objects::UClass*(*)(), bool);
+
+		int getVtableSize(void** vtable) {
+			if (vtable)
+				for (int i = 0;; i++)
+					if (!vtable[i])
+						return i;
+
+			return 0;
+		}
```
Comment (Vilsol) [inline SatisfactoryModLoader/assets/BPInterface.cpp]: Please follow the code guidelines by using brackets.

File: SatisfactoryModLoader/assets/BPInterface.h [hunk 3]
```diff
@@ -0,0 +1,1563 @@
...
+		SML_API extern void(*processNewObjs)();
+		SML_API extern void(*constructUClass)(Objects::UClass*&, FClassParams);
+		SML_API extern void(*privateStaticClassBody)(const TCHAR*, const TCHAR*, Objects::UClass*&, void(*)(), std::uint32_t, Objects::EClassFlags, Objects::EClassCastFlags, const TCHAR*, void(*)(FObjectInitializer&), Objects::UObject*(*)(FVtableHelper&), void(*)(Objects::UObject*, FReferenceCollector&), Objects::UClass*(*)(), Objects::UClass*(*)(), bool);
+
+		/**
+		* Converts the given normal string into a wide one
+		* ! WARNING ! No good implementation, does not support utf8! (only ASCII)
+		*
+		* @author Panakotta00
+		*/
```
Comment (Vilsol) [inline SatisfactoryModLoader/assets/BPInterface.h]: Please add a TODO comment regarding the warning.

File: SatisfactoryModLoader/util/Utility.cpp (hunk 1 above)
Comment (Panakotta00) [reply SatisfactoryModLoader/util/Utility.cpp]: Cause @SuperCoder7979 told me to try to make the new objects used for BPI independent from the SDK... I can try to change it though ^^

File: SatisfactoryModLoader/assets/BPInterface.cpp (hunk 2 above)
Comment (Panakotta00) [reply SatisfactoryModLoader/assets/BPInterface.cpp]: function made by @Brabb3l but I can change it ^^

File: SatisfactoryModLoader/util/Utility.cpp (hunk 1 above)
Comment (Vilsol) [reply SatisfactoryModLoader/util/Utility.cpp]: Nope, that's fine

**Pull request merged**